abi-versions/vNEXT/README.md

```diff
@@ -0,0 +1,1471 @@
+
+* params:
+  - `i32 (size_t) memory_size`
+* returns:
+  - `i32 (void*) memory_data`
+
+Allocates continuous memory buffer of size `memory_size` using the in-VM memory
+allocator and returns address of the memory buffer to the host environment using
+`memory_data`. Returning `0` indicates failure.
+
+
+## Context lifecycle
+
+### Callbacks optionally exposed by the Wasm module
+
+#### `proxy_on_context_create`
+
+* params:
+  - `i32 (uint32_t) context_id`
+  - `i32 (uint32_t) parent_context_id`
+  - `i32 (proxy_context_type_t) context_type`
+* returns:
+  - `i32 (uint32_t) new_context_id`
+
+Called when the host environment creates a new context (`context_id`) of type
+`context_type` (e.g. `VmContext`, `PluginContext`, `StreamContext`,
+`HttpContext`).
+
+This callback can be used to associate `context_id` with `parent_context_id`,
+or when using custom context identifiers, but otherwise the in-VM context can be
+implicitly created in `proxy_on_vm_start`, `proxy_on_plugin_start`,
+`proxy_on_new_connection` and `proxy_on_http_request_headers` callbacks.
+
+Return value (`new_context_id`) can be used to assign new context identifier
+that will be used by the host environment in subsequent calls. Returning `0`
+retains the original context identifier.
+
+#### `proxy_on_context_finalize`
+
+* params:
+  - `i32 (uint32_t) context_id`
+* returns:
+  - `i32 (bool) is_done`
+
+Called when the host environment is done processing context (`context_id`).
+Any finalization (e.g. logging) should be done within this callback.
+
+Return value (`is_done`) indicates that the Wasm module also finished processing
+context (`context_id`). If it didn't, then it should return `false` and call
+`proxy_context_finalize` when it's done.
+
+### Functions exposed by the host environment
+
+#### `proxy_set_effective_context`
+
+* params:
+  - `i32 (uint32_t) context_id`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Change the effective context to `context_id`. This function is usually used to
+change effective context during `proxy_on_queue_ready`, `proxy_on_timer_ready`,
+`proxy_on_http_call_response` and `proxy_on_grpc_call_response` callbacks.
+
+Returns `Ok` on success.
+
+Returns `BadArgument` for unknown `context_id`.
+
+#### `proxy_context_finalize`
+
+* params:
+  - none
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Indicates to the host environment that the Wasm module finished processing
+current context. It should be used after previously returning `false` in
+`proxy_on_context_finalize`.
+
+Returns `Ok` on success.
+
+
+## Configuration
+
+### Callbacks optionally exposed by the Wasm module
+
+#### `proxy_on_vm_start`
+
+* params:
+  - `i32 (uint32_t) vm_id`
+  - `i32 (size_t) vm_configuration_size`
+* returns:
+  - `i32 (bool) is_success`
+
+Called when the host environment starts WebAssembly Virtual Machine (WasmVm).
+
+Its configuration (of size `vm_configuration_size`) can be retrieved using
+`proxy_get_buffer`.
```

Comment:
FWIW I find this proxy_get_buffer dance to be a bit magical from an API design perspective. What if in the future we have two buffers that we want to get? IMO for this type of thing I think it would be much cleaner and more future proof to pass an opaque_handle that could then be passed back to proxy_get_buffer, or a size and handle if that makes more sense. This would make it very explicit what the user is fetching if we ever need to add the ability to fetch multiple things on the context of an API call.

Reply:
+1. I was going to comment on this, but I realized that for the practical use cases in the existing ABI that the key/value method seemed to work fine. But, I suspect that the proposal here will future proof and make the ABI a little more regular.